src/main/java/io/github/syst3ms/skriptparser/sections/SecChance.java

```diff
@@ -0,0 +1,70 @@
+package io.github.syst3ms.skriptparser.sections;
+
+import io.github.syst3ms.skriptparser.Main;
+import io.github.syst3ms.skriptparser.lang.CodeSection;
+import io.github.syst3ms.skriptparser.lang.Expression;
+import io.github.syst3ms.skriptparser.lang.Statement;
+import io.github.syst3ms.skriptparser.lang.TriggerContext;
+import io.github.syst3ms.skriptparser.parsing.ParseContext;
+import io.github.syst3ms.skriptparser.util.math.NumberMath;
+import org.jetbrains.annotations.Nullable;
+
+import java.math.BigDecimal;
+import java.util.Optional;
+import java.util.concurrent.ThreadLocalRandom;
+
+/**
+ * A condition that randomly succeeds or fails, given the chance for it to do succeed.
+ * Note that when the percent sign (%) is omitted, the chance is calculated from 0 to 1.
+ *
+ * @name Chance
+ * @type SECTION
+ * @pattern chance of %number%[\%]
+ * @since ALPHA
+ * @author Mwexim
+ */
+@SuppressWarnings("unchecked")
+public class SecChance extends CodeSection {
+
+    static {
+        Main.getMainRegistration().addSection(
+                SecChance.class,
+                "chance of %number%[1:\\%]"
+        );
+    }
+
+    private static final ThreadLocalRandom random = ThreadLocalRandom.current();
+
+    private Expression<Number> chance;
+    private boolean percent;
+
+    @Override
+    public boolean init(Expression<?>[] expressions, int matchedPattern, ParseContext parseContext) {
+        chance = (Expression<Number>) expressions[0];
+        percent = parseContext.getParseMark() == 1;
+        return true;
+    }
+
+    @Override
+    public Optional<? extends Statement> walk(TriggerContext ctx) {
+        Optional<? extends Number> c = chance.getSingle(ctx);
+        if (c.isEmpty())
+            return getNext();
+        double val = c.get().doubleValue();
+        if (val < 0 || val > (percent ? 100 : 1))
+            return getNext();
```

Comment:
This can be optimized using the Optional API, but since I was supposed to redo this logic myself anyway, I'll handle it myself while I'm at it.

Reply:
Yep, I left this one rather unchanged for you.